mongoose.c

```diff
@@ -3297,8 +3297,16 @@ static void http_cb(struct mg_connection *c, int ev, void *ev_data) {
         ofs += (size_t) n + hm.body.len;
       }
 
-      if (c->is_accepted) c->is_resp = 1;  // Start generating response
-      mg_call(c, MG_EV_HTTP_MSG, &hm);     // User handler can clear is_resp
+      if (c->is_accepted) {
+        struct mg_str *cc;
+        c->is_resp = 1;  // Start generating response
+        if ((cc = mg_http_get_header(&hm, "Connection")) != NULL &&
+            mg_strcasecmp(*cc, mg_str("close")) == 0) {
+          c->is_draining = 1;  // honor "Connection: close"
+          c->recv.len = 0;     // discard remaining requests, if any
+        }
+      }
+      mg_call(c, MG_EV_HTTP_MSG, &hm);  // User handler can clear is_resp
```

Comment:
This code changes the logic. Before, we were mg_call-ing always, for both client and server connections. Now, we ignore client connections. I guess we should keep the logic intact, by moving the mg_call outside of the `if` block.

I also would not clear `c->recv.len`, especially before calling the handler, as handler might use it.

A pedantic nitpick on the code - I suggest a slight refactoring:

```c
      if (c->is_accepted) {
        struct mg_str *cc = mg_http_get_header(&hm, "Connection");
        c->is_resp = 1;  // Start generating response
        if (cc != NULL && mg_strcasecmp(*cc, mg_str("close")) == 0) {
          c->is_draining = 1;  // honor "Connection: close"
        }
      }
```

Reply:
Nope, `mg_call()` was outside of the `if`. clang-format changed the comment spacing, so that line showed as modified.

Just setting c->is_draining is not enough, the user handler reply clears c->is_resp so the loop above continues and parses the next message:
https://github.com/cesanta/mongoose/blob/793171e9e0a36a5343313b53b98cf481ae2bf17f/src/http.c#L1006
Nevertheless, I see your point with clearing the buffer, got carried away by thinking pipelined requests are GETs and the message had already been parsed.
Refactored as suggested and placed it AFTER `mg_call`. Replaced buffer purge with a `break`, actually exiting the loop.

Reply:
Looks good! Approved.

Unit test please.

Reply:
> Unit test please

@cpq pls read comment on top, there's an issue on that. I could write a more complex test not using files but would need to add that one later.